File: src/hpc/run_lastz.pl
#!/usr/bin/env perl
#take a query fasta, and a bed
#the bee specifies expected mapping location
#run lastz using the expected chr to be mapped as ref, against a corresponding query sequence
#we use an entire chromsome as the reference, rather than a particular region on the chr
#such that we can maximize alignment

use strict;
use warnings;
use SeqMule::Utils;
use SeqMule::Parallel;
use File::Basename qw/basename/;
use File::Spec;

my $tmpdir = "/tmp";
my $desc = "
query_mapping.bed has 6 columns, query_ctg_id, start, end, chr, chr_start, chr_end
chr refers to the chromosome the query_ctg_id is mapped to
";
die "Usage: $0 <query.fa> [query_mapping.bed]\n" unless @ARGV == 2 or @ARGV == 1;
my $chr_dir = "/home/rcf-02/yunfeigu/proj_dir/pacbio_reference/data/hg38_chr/unmasked";
my $chrFile = "/home/rcf-02/yunfeigu/proj_dir/pacbio_reference/data/GCA_000001405.15_GRCh38_full_analysis_set.fna";
my $query = shift @ARGV;
my $bed = shift @ARGV;
my $debug = 1;
my %submitted;
my $qsubLaird = "qsub -V -l walltime=12:59:0 -l nodes=1:ppn=1 -A lc_kw -q laird -l mem=2GB -S /bin/bash";
my $qsubMain = "qsub -V -l walltime=12:59:0 -l nodes=1:ppn=1 -A lc_kw -l mem=2GB -S /bin/bash";
my $cwd = $ENV{PWD};
my $result_dir = File::Spec->catfile($cwd,"lastz_result");
mkdir $result_dir or die "mkdir($result_dir): $!\n" unless -d $result_dir;
if($bed) {
	open BED,'<',$bed or die "open($bed):$!\n";
	while (<BED>) {
		chomp;
		my @f = split;
		die "6 fields expected: $_ at $. of $bed\n" unless @f == 6;
		my ($qID,$qStart,$qEnd,$rID,$rStart,$rEnd) = @f;
		&submitLastzJob({qID=>$qID,rID=>$rID,result_dir=>$result_dir,query=>$query,chr_dir=>$chr_dir,qsub=>$qsubLaird});
	}
	close BED;
} else {
	#when no BED file, align each query contig to every chr, one at a time
	my %allContig = &SeqMule::Utils::readFastaIdx("$query.fai");
	my %allChr = &SeqMule::Utils::readFastaIdx("$chrFile.fai");
	my $count = 0;
	for my $i(keys %allContig) {
		for my $j(keys %allChr) {
			if(&getQCount('laird') <= 500) {
				&submitLastzJob({qID=>$i,rID=>$j,result_dir=>$result_dir,query=>$query,chr_dir=>$chr_dir,qsub=>$qsubLaird});
			} elsif (&getQCount('default')+&getQCount('main')+&getQCount('quick')+&getQCount('main_route') <= 300) {
				&submitLastzJob({qID=>$i,rID=>$j,result_dir=>$result_dir,query=>$query,chr_dir=>$chr_dir,qsub=>$qsubMain});
			} else {
				&blockOnQsub(
						{q=>['default','main','quick','main_route'],n=>300},
						{q=>['laird'],n=> 600}
					);
			}
		}
	}
}
warn "All done\n";

sub getQCount {
	#get number of jobs waiting on specified queue
	my $queueName = shift;
	my $count = `qstat -u yunfeigu|grep $queueName|wc -l`;
	chomp $count;
	warn "Jobs on $queueName: $count\n" if $debug;
	return $count;
}
sub blockOnQsub {
	#sleep until the jobs waiting on specified queues
	#go below certain threshold
	while(1) {
		sleep 1;
		for my $i(@_) {
			my $total = 0;
			map {$total += &getQCount($_)} @{$i->{q}};
			if ($total < $i->{n} ) {
				goto END;
			}
		}
	}
	END: 
	return;
}


=head
REF=/home/rcf-02/yunfeigu/proj_dir/pacbio_reference/data/hg38_chr/unmasked/chr1.fa
QUERY=/auto/rcf-proj/kw/yunfeigu/pacbio_reference/lastz/hx1f2a1_one_ctg_ref_hg38/000000F.fasta
lastz $REF $QUERY --notransition --step=20 --ambiguous=iupac --format=maf --gappedthresh=1000000 --identity=90 > one_ctg_on_hg38chr1.maf
=cut
#####################################################################
sub submitLastzJob {
	my $step = "
	1. mkdir invidual dir
	2. extract the query sequence
	3. generate script
	4. submit";
	my $opt = shift;
	#sleep int(rand(7));
	my ($qID,$rID,$result_dir,$query,$chr_dir,$qsub) = ($opt->{qID},$opt->{rID},$opt->{result_dir},$opt->{query},$opt->{chr_dir},$opt->{qsub});
	my $dir = File::Spec->catdir($result_dir,$qID);
	my $fa = File::Spec->catfile($dir,"query_$qID.fa");
	my $ref = File::Spec->catfile($dir,"ref_$rID.fa");
	my $stde = File::Spec->catfile($dir,"stderr");
	my $stdo = File::Spec->catfile($dir,"stdout");
	my $doneFile = File::Spec->catfile($dir,"$qID.$rID.done");
	mkdir $dir or die "mkdir($dir): $!\n" unless -d $dir;
	!system("samtools faidx $query $qID > $fa") or die "samtools faidx: $!\n";
	symlink File::Spec->catfile($chr_dir,"$rID.fa"),$ref or die "link(): $!\n" unless -e $ref or -l $ref;
	my $script = &SeqMule::Parallel::genTempScript(
		"cd \$TMPDIR",
		"lastz $ref $fa --notransition --step=20 --ambiguous=iupac --format=maf --gappedthresh=1000000 --identity=90 > query_${qID}_ref_${rID}.maf",
		"find . -name '*.maf' -size +4k | xargs -I{} cp {} $dir",
		"touch $doneFile");
	!system("$qsub -e $stde -o $stdo $script") or die "$script: $!\n" unless -e $doneFile or $submitted{$qID.$rID};
	#warn("$qsub -e $stde -o $stdo $script\n") or die "$script: $!\n" unless -e $doneFile or $submitted{$qID.$rID};
	$submitted{$qID.$rID} = 1;
}

File: src/assembly_comparison/hpc/run_lastz.py
#!/usr/bin/env python
'''take a query fasta, and optionally a bed
run lastz with hg38 as reference
if a bed is given, then each sequence in query fasta should be mapped a specific reference sequence
'''
import os
import sys
import logging
import tempfile
import time
logging.basicConfig(level=logging.INFO, format = '%(asctime)s - %(levelname)s - %(message)s')
#global variables
cleanQ = []
submitInterval = 1 #number of seconds between qstat
checkInterval = 1 #sec
debug = False
maxConcurrentLaird = 100
maxConcurrentBcl = 500
maxConcurrentOther = 100
jobReg = {} #record job queue, status, sentinel file path
#chr_dir = "/auto/rcf-proj/kw/yunfeigu/pacbio_reference/lastz/ref_hg38+decoy_query_hx1f4full_no_hg38+decoy.mummer/GRCh38_full_analysis_set_plus_decoy_hla.noMetaChar_chr"
#chrFile = "/auto/rcf-proj/kw/yunfeigu/pacbio_reference/lastz/ref_hg38+decoy_query_hx1f4full_no_hg38+decoy.mummer/GRCh38_full_analysis_set_plus_decoy_hla.noMetaChar.fa"
chr_dir = "/home/yunfeiguo/projects/PacBio_reference_genome/lastz/ref_novel_seq_absent_in_yh/chroms"
chrFile = "/home/yunfeiguo/projects/PacBio_reference_genome/lastz/ref_novel_seq_absent_in_yh/novel_by_ml_absent_in_yh.nometa.fasta"
#chrFile = "/home/yunfeiguo/database/hg_index/GRCh38_analysis/GRCh38_full_analysis_set_plus_decoy_hla.noMetaChar.fa"
#chr_dir = "/home/yunfeiguo/database/hg_index/GRCh38_analysis/GRCh38_full_analysis_set_plus_decoy_hla.noMetaChar_chr"
qsubLaird = "qsub -V -l walltime=199:59:0 -l nodes=1:ppn=1 -A lc_kw -q laird -l mem=2GB -S /bin/bash"
qsubMain = "qsub -V -l walltime=23:59:0 -l nodes=1:ppn=1 -A lc_kw -l mem=2GB -S /bin/bash"
qsubBcl = "qsub -V -cwd -l h_vmem=7G -S /bin/bash"
#prototype of lastz options
lastzOpt = "--notransition --gap=1000,1 --step=20 --ambiguous=iupac --format=maf --gappedthresh=10000 --identity=80 --progress=10 --maxwordcount=1 --masking=0 ";
cwd = os.getcwd()
#noSleep = False
noSleep = True

def safeMkdir(dir):
    if not os.path.isdir(dir):
        try:
            os.mkdir(dir)
        except OSError:
            logging.critical("%s exists" % dir)
def readFastaIdx(idx):
    fh = None
    try:
       fh = open(idx,'r')
    except IOError as e:
       logging.critical("I/O error: {0}".format(idx))
       raise
    ctg = {}
    for line in fh:
        f = line.split()
        assert len(f) == 5
        (id,length,offset,nCharLine,nByteLine) = f
        ctg[id] = {
                'length':length,
                'offset':offset,
                'nCharLine':nCharLine,
                'nByteLine':nByteLine,
                }
    fh.close()
    return(ctg)
def genTempScript(prefix,suffix,cmd):
    assert type(cmd) is list
    fd, tmpFile = tempfile.mkstemp(prefix = 'ref'+prefix, suffix='pairwiseLastz'+str(suffix))
    os.write(fd,'#!/bin/bash\n')
    os.write(fd,'set -e\n') #let shell run the script, exit at first error
    os.write(fd,'set -o pipefail\n') #let shell run the script, exit at first error
    for c in cmd:
        os.write(fd,c+'\n')
    os.close(fd)
    cleanQ.append(tmpFile) #clean at the very end
    return(tmpFile)
def clean():
    for f in cleanQ:
        os.remove(f)
def getJobCount(qName):
    '''
    check all jobs not finished
    remove any if finished
    return count for a particular queue
    '''
    #when all jobs are submitted, wait before check
    if not noSleep:
        if len(jobReg) == len([v for k,v in jobReg.items() if v['submitted']]):
            time.sleep(checkInterval)
            #pass
        else:
            time.sleep(submitInterval)
            #pass
    #make sure we mark a job as finished once it is done
    for i in jobReg:
        if jobReg[i]['finished']:
            pass
        else:
            if debug:
                print("checking %s" % jobReg[i]['doneFile'])
            if os.path.exists(jobReg[i]['doneFile']):
                if debug:
                    print("%s exists"%jobReg[i]['doneFile'])
                    print("%s is done"%i)
                jobReg[i]['finished'] = True
    total = len([v for k,v in jobReg.items() if v['submitted'] and v['q']==qName])
    finished = len([v for k,v in jobReg.items() if v['submitted'] and v['finished'] and v['q']==qName])
    if debug:
        print("total:%d,finished:%d"%(total,finished))
    return(total-finished)

def submitLastzJob(result_dir, fa, qCount, rID, q, qsub):
    '''
    example lastz command
    REF=/home/rcf-02/yunfeigu/proj_dir/pacbio_reference/data/hg38_chr/unmasked/chr1.fa
    QUERY=/auto/rcf-proj/kw/yunfeigu/pacbio_reference/lastz/hx1f2a1_one_ctg_ref_hg38/000000F.fasta
    lastz $REF $QUERY --notransition --step=20 --ambiguous=iupac --format=maf --gappedthresh=10000 --identity=90 > one_ctg_on_hg38chr1.maf
    
    steps to submit
    	1. mkdir invidual dir
    	2. extract the query sequence
    	3. generate script
    	4. submit"
    '''
    if not noSleep:
        time.sleep(submitInterval)
    dir = os.path.join(result_dir,rID)
    ref = os.path.join(dir,"ref_%s.fa" % rID)
    stde = os.path.join(dir,"stderr"+str(qCount))
    stdo = os.path.join(dir,"stdout"+str(qCount))
    jobID = rID + str(qCount)
    if debug:
        print(jobID)
    individualDoneFile = os.path.join(dir,"%s_query%d.done" % (rID,qCount))
    safeMkdir(dir)
    jobReg[jobID] = {
            'doneFile':individualDoneFile,
            'q':q,
            'submitted':False,
            'finished':False,
            }
    script = None
    if os.path.lexists(ref):
        os.remove(ref)
    #create the symlink for reference regardless it existence
    os.symlink(os.path.join(chr_dir,"%s.fa"%rID),ref)
    script = genTempScript(rID, qCount, [
        "cd $TMPDIR",
		"lastz %s %s %s > ref_%s_vs_query%d.maf" % (ref,fa,lastzOpt,rID,qCount),
		"find . -name '*.maf' -size +4k | xargs -I{} cp {} %s" % (dir),
		"touch %s" % (individualDoneFile),
        ])
    if script is None:
        logging.critical("Nonetype for script")
    if debug:
        print("done file exists?")
        print(os.path.exists(individualDoneFile))
        print("submitted?")
        print(jobReg[jobID]['submitted'])
    if os.path.exists(individualDoneFile) or jobReg[jobID]['submitted']: 
        pass
    else:
        os.system("%s -e %s -o %s %s" % (qsub, stde, stdo, script))
        #print("%s -e %s -o %s %s" % (qsub, stde, stdo, script))
        jobReg[jobID]['submitted'] = True
#####################################################################
desc = '''
query_mapping.bed has 6 columns, query_ctg_id, start, end, chr, chr_start, chr_end
chr refers to the chromosome the query_ctg_id is mapped to'''
if len(sys.argv) == 0 or len(sys.argv) == 1:
    sys.exit("Usage: %s <query1.fa query2.fa ...>" % sys.argv[0])
query = sys.argv[1:len(sys.argv)]
bed = None
result_dir = os.path.join(cwd,"lastz_result")
safeMkdir(result_dir)
	#align each query contig to every chr, one at a time
#allChr = ["alt","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr1","chr20","chr21","chr22","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chrX","chrY"]
allChr = readFastaIdx(chrFile+".fai")
for j in range(0,len(query)):
    oneQuery = os.path.abspath(query[j])
    print(oneQuery)
    for i in allChr:
        #if getJobCount('laird') < maxConcurrentLaird:
        #    submitLastzJob(result_dir = result_dir, fa = oneQuery, qCount = j, rID = i, q = 'laird', qsub = qsubLaird)
        #elif getJobCount('other') < maxConcurrentOther:
        #    submitLastzJob(result_dir = result_dir, fa = oneQuery, qCount = j, rID = i, q = 'other', qsub = qsubMain)
        if getJobCount('all.q') < maxConcurrentBcl:
            submitLastzJob(result_dir = result_dir, fa = oneQuery, qCount = j, rID = i, q = 'all.q', qsub = qsubBcl)
        else:
            if debug:
                print("queuing")
            while True:
		if getJobCount('all.q') < maxConcurrentBcl:
                #if getJobCount('laird') < maxConcurrentLaird or getJobCount('other') < maxConcurrentOther:
                    break
#clean()
logging.info("All done")
